services-api/src/main/java/io/scalecube/services/Reflect.java

```diff
@@ -270,7 +297,14 @@ public static void validateMethodOrThrow(Method method) {
       throw new UnsupportedOperationException("Service method return type can be Publisher only");
     }
     if (method.getParameterCount() > 1) {
-      throw new UnsupportedOperationException("Service method can accept 0 or 1 parameters only");
+      int paramsCount = 0;
+
+      for (Parameter parameter : method.getParameters()) {
+        if (!parameter.isAnnotationPresent(Principal.class) && ++paramsCount > 1) {
```

Comment:
Can we do it without `arameter.isAnnotationPresent(Principal.class)`  in the critical path ?

Reply:
we should cache this (as method signature is *Hard Coded*)

Reply:
Implementation is changed a bit, but still contains `parameter.isAnnotationPresent(Principal.class)`.

Anyway, `validateMethodOrThrow` is not the critical path, since it is called only on service registration.